File: patterns/structural/proxy.py
"""
Proxy Pattern - Python

Provides a placeholder or surrogate for another object to control access to it.
Acts as an intermediary between client and the real object.

Use cases:
- Lazy loading
- Access control
- Logging
- Caching
- Remote proxies
"""

from abc import ABC, abstractmethod


# Subject interface
class Image(ABC):
    @abstractmethod
    def display(self):
        pass


# Real subject
class RealImage(Image):
    def __init__(self, filename):
        self.filename = filename
        self.load_from_disk()
    
    def load_from_disk(self):
        print(f"Loading {self.filename} from disk...")
    
    def display(self):
        return f"Displaying {self.filename}"


# Proxy
class ProxyImage(Image):
    def __init__(self, filename):
        self.filename = filename
        self.real_image = None
    
    def display(self):
        if not self.real_image:
            self.real_image = RealImage(self.filename)
        return self.real_image.display()


# Another example: Protection Proxy
class BankAccount:
    def __init__(self, balance):
        self.balance = balance
    
    def withdraw(self, amount):
        if amount > self.balance:
            raise ValueError("Insufficient funds")
        self.balance -= amount
        return f"Withdrew ${amount}. New balance: ${self.balance}"
    
    def get_balance(self):
        return self.balance


class BankAccountProxy:
    def __init__(self, account, user):
        self.account = account
        self.user = user
    
    def withdraw(self, amount):
        if self.user != 'admin':
            if amount > 1000:
                raise ValueError("Non-admin users cannot withdraw more than $1000")
        return self.account.withdraw(amount)
    
    def get_balance(self):
        return self.account.get_balance()


# Example usage
if __name__ == "__main__":
    print("=== Proxy Pattern ===\n")
    
    # Virtual Proxy (Lazy Loading)
    print("Virtual Proxy (Lazy Loading):")
    image1 = ProxyImage('photo1.jpg')
    image2 = ProxyImage('photo2.jpg')
    
    # Image not loaded yet
    print("Image created, but not loaded")
    
    # Now load and display
    print(image1.display())
    print(image2.display())
    
    print("\nProtection Proxy:")
    account = BankAccount(5000)
    user_proxy = BankAccountProxy(account, 'user')
    admin_proxy = BankAccountProxy(account, 'admin')
    
    print("User balance:", user_proxy.get_balance())
    print(user_proxy.withdraw(500))
    try:
        print(user_proxy.withdraw(2000))  # Should fail
    except ValueError as e:
        print("Error:", e)
    
    print("Admin can withdraw any amount:")
    print(admin_proxy.withdraw(2000))

File: patterns/structural/proxy.js
/**
 * Proxy Pattern - JavaScript
 * 
 * Provides a placeholder or surrogate for another object to control access to it.
 * Acts as an intermediary between client and the real object.
 * 
 * Use cases:
 * - Lazy loading
 * - Access control
 * - Logging
 * - Caching
 * - Remote proxies
 */

// Subject interface
class Image {
    display() {
        throw new Error('Method must be implemented');
    }
}

// Real subject
class RealImage extends Image {
    constructor(filename) {
        super();
        this.filename = filename;
        this.loadFromDisk();
    }
    
    loadFromDisk() {
        console.log(`Loading ${this.filename} from disk...`);
    }
    
    display() {
        return `Displaying ${this.filename}`;
    }
}

// Proxy
class ProxyImage extends Image {
    constructor(filename) {
        super();
        this.filename = filename;
        this.realImage = null;
    }
    
    display() {
        if (!this.realImage) {
            this.realImage = new RealImage(this.filename);
        }
        return this.realImage.display();
    }
}

// Another example: Protection Proxy
class BankAccount {
    constructor(balance) {
        this.balance = balance;
    }
    
    withdraw(amount) {
        if (amount > this.balance) {
            throw new Error('Insufficient funds');
        }
        this.balance -= amount;
        return `Withdrew $${amount}. New balance: $${this.balance}`;
    }
    
    getBalance() {
        return this.balance;
    }
}

class BankAccountProxy {
    constructor(account, user) {
        this.account = account;
        this.user = user;
    }
    
    withdraw(amount) {
        if (this.user !== 'admin') {
            if (amount > 1000) {
                throw new Error('Non-admin users cannot withdraw more than $1000');
            }
        }
        return this.account.withdraw(amount);
    }
    
    getBalance() {
        return this.account.getBalance();
    }
}

// Example usage
console.log('=== Proxy Pattern ===\n');

// Virtual Proxy (Lazy Loading)
console.log('Virtual Proxy (Lazy Loading):');
const image1 = new ProxyImage('photo1.jpg');
const image2 = new ProxyImage('photo2.jpg');

// Image not loaded yet
console.log('Image created, but not loaded');

// Now load and display
console.log(image1.display());
console.log(image2.display());

console.log('\nProtection Proxy:');
const account = new BankAccount(5000);
const userProxy = new BankAccountProxy(account, 'user');
const adminProxy = new BankAccountProxy(account, 'admin');

console.log('User balance:', userProxy.getBalance());
console.log(userProxy.withdraw(500));
try {
    console.log(userProxy.withdraw(2000)); // Should fail
} catch (e) {
    console.log('Error:', e.message);
}

console.log('Admin can withdraw any amount:');
console.log(adminProxy.withdraw(2000));

module.exports = { Image, RealImage, ProxyImage, BankAccount, BankAccountProxy };

